### DISCUSSION TITLE
Uniform drift rate variability always zero

### DISCUSSION THREAD

**Discussion by Jiashun97** [mwshinn/PyDDM#88] [Ask for help]
Hi Max,

In my analysis, there are two speed-accuracy trade-off conditions (A and B), so the "DriftUniformSAT" class has three parameters. However, the results showed that all width parameter values are zeros. I was wondering if you have also encountered similar situations. I have implemented the intertrial variability for drift rate below. It would be great if you can share some insights.

<img width="1000" alt="Screenshot 2024-01-16 at 16 14 35" src="https://github.com/mwshinn/PyDDM/assets/65221219/fe33dd09-61fe-4f37-9c21-323f9d4d19ba">

`def prepare_sample_for_variable_drift(sample, resolution=11):
    new_samples = []
    for i in range(0, resolution):
        choice_upper = sample.choice_upper.copy()
        choice_lower = sample.choice_lower.copy()
        undecided = sample.undecided
        conditions = copy.deepcopy(sample.conditions)
        conditions['driftnum'] = (np.asarray([i]*len(choice_upper)),
                                  np.asarray([i]*len(choice_lower)),
                                  np.asarray([i]*undecided))
        new_samples.append(Sample(choice_upper, choice_lower, undecided, choice_names=sample.choice_names, **conditions))
    new_sample = new_samples.pop()
    for s in new_samples:
        new_sample += s
    return new_sample`

`class DriftUniformSAT(Drift):
    name = "Uniformly-distributed drift"
    resolution = 11 # Number of bins, should be an odd number
    required_parameters = ['Vspeed', 'Vacc', 'width'] # Mean drift and the width of the uniform distribution
    required_conditions = ['driftnum']
    def get_drift(self, x, t, conditions, **kwargs):
        if conditions['cond'] == 'A':
            stepsize = self.width/(self.resolution-1)
            mindrift = self.Vspeed - self.width/2
            return mindrift + stepsize*conditions['driftnum']
        elif conditions['cond'] == 'B':
            stepsize = self.width/(self.resolution-1)
            mindrift = self.Vacc - self.width/2
            return mindrift + stepsize*conditions['driftnum']
        else: 
            print("Non-implemented SAT conditions")`

Best regards,
Jiashun

**Comment by mwshinn** [comment]
Please post a minimal example script that shows the full problem in as simple of a way as possible.  Something that can be run verbatim by copying and pasting.  Also, see if you can figure out if the model pdf changes if you manually fix the width parameter and then solve (i.e., is this parameter actually doing something in your model).

**Reply by Jiashun97** [reply]
[pp1.txt](https://github.com/mwshinn/PyDDM/files/13962089/pp1.txt)
[demo.ipynb.zip](https://github.com/mwshinn/PyDDM/files/13962106/demo.ipynb.zip)

Thanks for your quick reply. I created an example of one participant with two conditions. The v, a, and ter are free to vary across condition, but the sv is fixed.

**Reply by mwshinn** [reply]
Sorry, if I wasn't clear, by "simple as possible" I meant strip away all of the parts that are not necessary for reproducing the issue, e.g., if you still get the issue when you remove fittable bounds, then remove fittable bounds and just use constant bounds.  Also see if you can get to the root of the problem just using the "solve" method without actually fitting the data or even using data at all.  You can also try minimising the amount of discretisation in the drift distribution.  And take the model fitting code out of the function and just pass values directly.  The reason I suggest all of this is that often times you will discover the source of the problem by going through this process of finding the minimal example.  If not, I can take another look at the minimal example you come up with.  I didn't see anything wrong with the drift rate variability code you sent at first glance.

**Reply by Jiashun97** [reply]
Thanks for your suggestion! I'll try the solution to see if it works.
